src/front/wgsl/parse/ast.rs

```diff
@@ -0,0 +1,284 @@
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum GlobalDeclKind {
+    Fn(Fn),
+    Override(Override),
+    Var(Var),
+    Let(Let),
+    Const(Let),
+    StaticAssert(StaticAssert),
+    Struct(Struct),
+    Type(TypeDecl),
+}
+
+#[derive(Clone, Debug)]
+pub struct Fn {
+    pub attribs: Vec<Attribute>,
+    pub name: Ident,
+    pub args: Vec<Arg>,
+    pub ret_attribs: Vec<Attribute>,
+    pub ret: Option<Type>,
+    pub block: Block,
+}
+
+#[derive(Clone, Debug)]
+pub struct Override {
+    pub attribs: Vec<Attribute>,
+    pub name: Ident,
+    pub ty: Option<Type>,
+    pub val: Option<Expr>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Var {
+    pub attribs: Vec<Attribute>,
+    pub inner: VarNoAttribs,
+}
+
+#[derive(Clone, Debug)]
+pub struct VarNoAttribs {
+    pub address_space: Option<Ident>,
+    pub access_mode: Option<Ident>,
+    pub name: Ident,
+    pub ty: Option<Type>,
+    pub val: Option<Expr>,
+}
+
+#[derive(Clone, Debug)]
+pub struct StaticAssert {
+    pub expr: Expr,
+}
+
+#[derive(Clone, Debug)]
+pub struct Struct {
+    pub name: Ident,
+    pub fields: Vec<Arg>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TypeDecl {
+    pub name: Ident,
+    pub ty: Type,
+}
+
+#[derive(Clone, Debug)]
+pub struct Arg {
+    pub attribs: Vec<Attribute>,
+    pub name: Ident,
+    pub ty: Type,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub struct Type {
+    pub kind: TypeKind,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum TypeKind {
+    Ident(Ident, Vec<Type>),
+    Array(Ident, Box<Type>, Option<Expr>),
+}
+
+#[derive(Clone, Debug)]
+pub struct Block {
+    pub stmts: Vec<Stmt>,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub struct Stmt {
+    pub kind: StmtKind,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum StmtKind {
+    Block(Block),
+    Expr(Expr),
+    Break,
+    Continue,
+    Discard,
+    For(For),
+    If(If),
+    Loop(Block),
+    Return(Option<Expr>),
+    StaticAssert(StaticAssert),
+    Switch(Switch),
+    While(While),
+    Continuing(Block),
+    BreakIf(Expr),
+    Empty,
+}
+
+#[derive(Clone, Debug)]
+pub struct For {
+    pub init: Option<Expr>,
+    pub cond: Option<Expr>,
+    pub update: Option<Expr>,
+    pub block: Block,
+}
+
+#[derive(Clone, Debug)]
+pub struct If {
+    pub cond: Expr,
+    pub block: Block,
+    pub else_: Option<Box<Stmt>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct While {
+    pub cond: Expr,
+    pub block: Block,
+}
+
+#[derive(Clone, Debug)]
+pub struct Switch {
+    pub expr: Expr,
+    pub cases: Vec<Case>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Case {
+    pub selectors: Vec<CaseSelector>,
+    pub block: Block,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum CaseSelector {
+    Expr(Expr),
+    Default,
+}
+
+#[derive(Clone, Debug)]
+pub struct Expr {
+    pub kind: ExprKind,
+    pub span: Span,
+}
+
+#[derive(Clone, Debug)]
+pub enum ExprKind {
+    Underscore,
+    VarDecl(Box<VarDecl>),
```

Comment:
It is very surprising to see variable declarations as an expression type. Overall, I had a hard time tracking how identifiers were resolved. I think this could be simplified a lot, while still implementing WGSL's semantics.